### PR TITLE
Add a config option to enable Flyway baseline feature

### LINKED ISSUE #27734: flyway baseline-version has no effect
### Describe the bug

After updating the config key quarkus.flyway.baseline-version: "2022.09.02.2", I was expecting newly created environments to use 2022.09.02.2 as first version (with migrate-at-start enabled). 

It turns out that migrate-at-start does not execute the baseline operation and all migrations before 2022.09.02.2 are still executed.

### Expected behavior

first migration of newly created environments is equal to baseline-version

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.12.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by rmanibus**
Resolves #27734

### CODE REVIEW

**Comment by geoand** [comment]
Thanks for this.

It seems like the test does not compile

**Comment by rmanibus** [comment]
@geoand I was struggling to get the project imported correctly by intellij, it should be good now.

I also rephrased the baselineOnMigrate config doc because the current description is a bit misleading

**Comment by rmanibus** [comment]
@geoand weird that FlywayExtensionMigrateAtStartTest is failing. H2 is not reset between tests ?

**Comment by geoand** [comment]
I'm pretty sure it's not

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java [hunk 1]
```diff
@@ -69,6 +70,10 @@ public Flyway get() {
 
     public void doStartActions() {
         for (FlywayContainer flywayContainer : FLYWAY_CONTAINERS) {
+            if (flywayContainer.isBaselineAtStart() &&
+                    Objects.isNull(flywayContainer.getFlyway().info().current())) {
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:74]:
```suggestion
            if (flywayContainer.isBaselineAtStart() &&
                    flywayContainer.getFlyway().info().current() == null) {
```

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java [hunk 2]
```diff
@@ -69,6 +70,10 @@ public Flyway get() {
...
+                flywayContainer.getFlyway().baseline();
+            }
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:76]: I think we should analyze the order on how this change affects the others `*AtStart` flags. 
For example: 
- Do we have any mutual exclusion? 
- Should `clean()` be always executed first?

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java [hunk 3]
```diff
@@ -113,11 +113,20 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
...
-     * Enable the creation of the history table if it does not exist already.
+     * call baseline when migrate is executed against a non-empty schema with no metadata table.
+     * This schema will then be baselined with the baselineVersion before executing the migrations.
+     * Only migrations above baselineVersion will then be applied
      */
     @ConfigItem
     public boolean baselineOnMigrate;
 
+    /**
+     * true to execute Flyway baseline automatically when the application starts when executed against an empty schema, false
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:124]: Apparently this command fails if the `flyway_schema_history` table exists (see https://github.com/quarkusio/quarkus/issues/27734#issuecomment-1238813838).

Comment (rmanibus) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:124]: I will add a test with an existing table to be sure

**Comment by gastaldi** [comment]
> @geoand weird that FlywayExtensionMigrateAtStartTest is failing. H2 is not reset between tests ?

Take a look at how it is done in https://github.com/quarkusio/quarkus/blob/main/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionCallbackTest.java

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java (hunk 3 above)
Comment (rmanibus) [reply extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:124]: @gastaldi , I made some test and here is my conclusion:

if the flyway_schema_history table exist and contains any element, everything is ok.
if the flyway_schema_history does not exist, everything is ok.
if the flyway_schema_history exist but is empty, then flyway.info().current() is null so we do execute the baseline and it is failing. 

I will try to find a way to check if flyway_schema_history already exist in flyway api.

**Comment by rmanibus** [comment]
> For the record, I'm running full Quarkus builds locally (completely unrelated to this PR) and I'm seeing `FlywayExtensionMigrateAtStartTest` fail as well. So I believe that's not caused by this PR, but sneaked in through some other PR.

@Ladicek I just changed the DB name in the connection string of my test and everything is working fine now

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java [hunk 4]
```diff
@@ -72,6 +73,15 @@ public void doStartActions() {
...
                 flywayContainer.getFlyway().clean();
             }
+            if (flywayContainer.isBaselineAtStart()) {
+                new FlywayExecutor(flywayContainer.getFlyway().getConfiguration())
+                        .execute((m, schemaHistory, d, ds, s, ce, si) -> {
+                            if (!schemaHistory.exists()) {
+                                return flywayContainer.getFlyway().baseline();
+                            }
+                            return null;
+                        }, true);
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:83]: Please avoid lambdas in the recorder code

**Comment by gsmet** [comment]
@gastaldi can you have another look?

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java (hunk 3 above)
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:124]:
```suggestion
     * true to execute Flyway baseline automatically when the application starts. This flag is ignored if the flyway_schema_history table does not exist in the current schema.
```

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java [hunk 5]
```diff
@@ -113,11 +113,20 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
     public boolean repairAtStart;
 
     /**
-     * Enable the creation of the history table if it does not exist already.
+     * call baseline when migrate is executed against a non-empty schema with no metadata table.
+     * This schema will then be baselined with the baselineVersion before executing the migrations.
+     * Only migrations above baselineVersion will then be applied
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:118]: This is confusing, can you reword it?

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java [hunk 6]
```diff
@@ -80,4 +92,21 @@ public void doStartActions() {
...
+        BaselineCommand(Flyway flyway) {
+            this.flyway = flyway;
+        }
+
+        final Flyway flyway;
+
+        @Override
+        public BaselineResult execute(CompositeMigrationResolver cmr, SchemaHistory schemaHistory, Database d,
+                Schema defaultSchema, Schema[] s, CallbackExecutor ce, StatementInterceptor si) {
+            if (!schemaHistory.exists()) {
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java:106]: You said previously:
> if the flyway_schema_history table exist and contains any element, everything is ok.

Why just checking if the `flyway_schema_history` does not exist here?

**Review by gastaldi — CHANGES_REQUESTED** [review]
I added some comments. I think it still needs some clarification

**Comment by geoand** [comment]
This is going to need a rebase

**Comment by rmanibus** [comment]
@geoand I rebased it

**Comment by geoand** [comment]
Thanks!

Have the comments made during the review been addressed? cc @gastaldi

**Comment by rmanibus** [comment]
If I remember well I addressed all the concerns.
I could also include a guide on how to perform migrations squashing in the documentation.

File: extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java [hunk 7]
```diff
@@ -141,11 +141,22 @@ public static FlywayDataSourceRuntimeConfig defaultConfig() {
...
+     * current schema or if the current schema is empty.
+     * Note that this will not automatically call migrate, you must either enable baselineAtStart or programmatically call
+     * flyway.migrate().
      */
     @ConfigItem
     public boolean baselineOnMigrate;
 
+    /**
+     * true to execute Flyway baseline automatically when the application starts.
+     * This flag is ignored if the flyway_schema_history table exist in the current schema.
```
Comment (gastaldi) [inline extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java:154]:
```suggestion
     * This flag is ignored if the flyway_schema_history table exists in the current schema.
```

**Review by gastaldi — APPROVED** [review]
LGTM, one minor observation

**Comment by rmanibus** [comment]
> If I remember well I addressed all the concerns. I could also include a guide on how to perform migrations squashing in the documentation.

fixed the typo. I will do this in another PR

**Comment by gastaldi** [comment]
Could you squash the commits before we trigger CI?

**Comment by Sanne** [comment]
@gastaldi ping ? This looks like ready but it's waiting for your opinion

**Comment by gastaldi** [comment]
@Sanne thanks, I've rebased the branch, it should be good to merge once it passes

**Comment by gastaldi** [comment]
@rmanibus because I forgot to merge this PR 3 months ago, this now needs to be changed to match the latest Flyway codebase. Can you have a look?

**Comment by rmanibus** [comment]
@gastaldi it is rebased !

**Comment by gastaldi** [comment]
Can you rebase and remove the merge commit?

**Comment by rmanibus** [comment]
@gastaldi done !

**Pull request merged**